File: backend/app/downloader.py
import re
import uuid
import yt_dlp
import yt_dlp.extractor as extractors
from .db import get_db
from .config import MAX_VIDEO_DURATION_SEC, DOWNLOAD_DIR

def clean(msg: str):
    return re.sub(r'\x1B\[[0-?]*[ -/]*[@-~]', '', msg)

def validate_url(url: str):
    for ie in extractors.gen_extractors():
        if ie.suitable(url):
            return
    raise Exception("URL not supported by yt-dlp")

def get_download_options(url: str):
    validate_url(url)

    with yt_dlp.YoutubeDL({"quiet": True, "skip_download": True}) as ydl:
        try:
            info = ydl.extract_info(url, download=False)
        except Exception as e:
            raise Exception(clean(str(e)))

    if info.get("duration", 0) > MAX_VIDEO_DURATION_SEC:
        raise Exception("Video duration exceeds limit")

    options_id = uuid.uuid4().hex
    db = get_db()

    response = {
        "options_id": options_id,
        "title": info.get("title"),
        "duration": info.get("duration"),
        "video_audio": [],
        "video_only": [],
        "audio": []
    }

    seen_ids = set()
    for f in info.get("formats", []):
        fid = f.get("format_id")
        v, a = f.get("vcodec"), f.get("acodec")
        h, abr = f.get("height"), f.get("abr")

        if v != "none" and a != "none" and h:
            key = f"va_{h}"
            label = f"{h}p"
            category = "video_audio"
        elif v != "none" and a == "none" and h:
            key = f"vo_{h}"
            label = f"{h}p"
            category = "video_only"
        elif v == "none" and a != "none":
            key = f"a_{int(abr or 0)}"
            label = f"{int(abr or 0)} kbps"
            category = "audio"
        else:
            continue
        
        if key in seen_ids:
            continue
        seen_ids.add(key)
        
        response[category].append({"id": key, "label": label})

        db.execute(
            "INSERT OR IGNORE INTO options VALUES (?, ?, ?)",
            (options_id, key, fid)
        )

    db.commit()
    db.close()
    return response

def download_video(job_id: str, url: str, format_id: str):
    import os

    def hook(d):
        db = get_db()
        # Check for cancellation
        status_row = db.execute("SELECT status FROM jobs WHERE id=?", (job_id,)).fetchone()
        if status_row and status_row["status"] == "cancelled":
            db.close()
            raise Exception("Download cancelled by user")

        if d["status"] == "downloading":
             db.execute(
                "UPDATE jobs SET status=?, progress=?, speed=?, eta=? WHERE id=?",
                (
                    "downloading", 
                    clean(d.get("_percent_str", "0%")), 
                    clean(d.get("_speed_str", "N/A")), 
                    clean(d.get("_eta_str", "N/A")), 
                    job_id
                )
            )
        db.commit()
        db.close()

    ydl_opts = {
        "format": format_id,
        "outtmpl": f"{DOWNLOAD_DIR}/{job_id}/%(title).200s [%(id)s].%(ext)s",
        "restrictfilenames": True,
        "windowsfilenames": True,
        "progress_hooks": [hook],
        "quiet": True,
        "noplaylist": True,
        "merge_output_format": "mp4",
        "no_continue": True,
        "overwrites": True
    }

    try:
        with yt_dlp.YoutubeDL(ydl_opts) as ydl:
            # 1. Get info to calculate filename
            info = ydl.extract_info(url, download=False)
            
            # 2. Calculate final filename (forcing mp4 due to merge_output_format)
            # Note: prepare_filename returns full path, we need relative to DOWNLOAD_DIR
            full_path = ydl.prepare_filename(info)
            filename = os.path.basename(full_path)
            base, _ = os.path.splitext(filename)
            
            # Storing relative path: job_id/filename.mp4
            final_filename = f"{job_id}/{base}.mp4"

            # 3. Perform download
            ydl.download([url])
            
            # 4. Update DB only after success
            db = get_db()
            db.execute(
                "UPDATE jobs SET status=?, progress=?, filename=? WHERE id=?",
                ("finished", "100%", final_filename, job_id)
            )
            db.commit()
            db.close()
            
    except Exception as e:
        msg = clean(str(e))
        db = get_db()
        # Don't overwrite cancelled status with error
        current = db.execute("SELECT status FROM jobs WHERE id=?", (job_id,)).fetchone()
        if current and current["status"] != "cancelled":
             db.execute(
                "UPDATE jobs SET status=?, error=? WHERE id=?",
                ("error", msg, job_id)
            )
        db.commit()
        db.close()

File: backend-node/downloader.js
'use strict';
const { spawn }  = require('child_process');
const { randomBytes } = require('crypto');
const fs   = require('fs');
const path = require('path');
const config = require('./config');
const { getDb } = require('./db');

const _activeProcs = new Map();
const _PROGRESS_RE = /\[PROG\](\w+)\|([^|]+)\|([^|]+)\|([^\n]+)/;

// ── Helpers ──────────────────────────────────────────────────────────────────

function clean(msg) {
  return (msg || '')
    .replace(/\x1B\[[0-?]*[ -/]*[@-~]/g, '')   // ANSI codes
    .replace(/^ERROR:\s*/gm, '')                  // yt-dlp "ERROR:" prefix
    .split(/;\s*please report this issue/i)[0]
    .trim();
}

function validateUrl(url) {
  if (!url || url.startsWith('-')) throw new Error('Invalid URL');
  let parsed;
  try { parsed = new URL(url); } catch { throw new Error('Invalid URL'); }
  if (!['http:', 'https:'].includes(parsed.protocol)) {
    throw new Error('Only http/https URLs are supported');
  }
}

function generateId() {
  return randomBytes(16).toString('hex');
}

function baseCmd() {
  const cmd = [config.YTDLP_BINARY];
  if (config.NODE_BINARY && fs.existsSync(config.NODE_BINARY)) {
    cmd.push('--js-runtimes', `node:${config.NODE_BINARY}`);
  }
  return cmd;
}

function runToCompletion(exe, args, timeoutMs = 120_000) {
  return new Promise((resolve) => {
    const proc = spawn(exe, args, { timeout: timeoutMs });
    let stdout = '', stderr = '';
    proc.stdout.on('data', d => { stdout += d; });
    proc.stderr.on('data', d => { stderr += d; });
    proc.on('close', code => resolve({ code, stdout, stderr }));
    proc.on('error', err => resolve({ code: 1, stdout: '', stderr: err.message }));
  });
}

// ── Public API ────────────────────────────────────────────────────────────────

async function getDownloadOptions(url) {
  validateUrl(url);

  const [exe, ...args] = [...baseCmd(), '--dump-json', '--no-playlist', '--quiet', '--no-warnings', '--', url];
  const { code, stdout, stderr } = await runToCompletion(exe, args);

  if (code !== 0) throw new Error(clean(stderr || stdout) || 'Failed to fetch video info');

  const raw = stdout.trim();
  if (!raw) throw new Error('No video info returned — URL may not be supported');

  let info;
  try { info = JSON.parse(raw); } catch { throw new Error('Failed to parse video info'); }


  const optionsId = generateId();
  const db = getDb();
  const insertOpt = db.prepare('INSERT OR IGNORE INTO options VALUES (?, ?, ?)');

  const thumbnail =
    info.thumbnail ||
    (Array.isArray(info.thumbnails) && info.thumbnails.length
      ? info.thumbnails[info.thumbnails.length - 1].url
      : null);

  const response = {
    options_id: optionsId,
    title:       info.title,
    duration:    info.duration,
    thumbnail,
    video_audio: [],
    video_only:  [],
    audio:       [],
  };

  const seen = new Set();
  for (const f of (info.formats || [])) {
    const { format_id: fid, vcodec, acodec, height, abr } = f;
    let key, label, category;

    if (vcodec !== 'none' && acodec !== 'none' && height) {
      key = `va_${height}`;  label = `${height}p`;                    category = 'video_audio';
    } else if (vcodec !== 'none' && acodec === 'none' && height) {
      key = `vo_${height}`;  label = `${height}p`;                    category = 'video_only';
    } else if (vcodec === 'none' && acodec !== 'none') {
      const kbps = Math.round(abr || 0);
      key = `a_${kbps}`;     label = `${kbps} kbps`;                 category = 'audio';
    } else {
      continue;
    }

    if (seen.has(key)) continue;
    seen.add(key);

    response[category].push({ id: key, label });
    insertOpt.run(optionsId, key, fid);
  }

  return response;
}

// Runs in the background — never awaited by the request handler.
async function downloadVideo(jobId, url, formatId) {
  if (!formatId || formatId.startsWith('-')) throw new Error(`Invalid formatId: ${formatId}`);
  validateUrl(url);

  const outdir = path.join(config.DOWNLOAD_DIR, jobId);
  fs.mkdirSync(outdir, { recursive: true });

  const outtmpl = path.join(outdir, '%(title).200s [%(id)s].%(ext)s');

  const [exe, ...args] = [
    ...baseCmd(),
    '--format',            formatId,
    '--output',            outtmpl,
    '--newline',
    '--no-playlist',
    '--restrict-filenames',
    '--windows-filenames',
    '--merge-output-format', 'mp4',
    '--no-continue',
    '--force-overwrites',
    '--ffmpeg-location',   config.FFMPEG_LOCATION,
    '--cache-dir',         config.TEMP_DIR,
    '--no-warnings',
    '--progress-template', '[PROG]%(progress.status)s|%(progress._percent_str)s|%(progress._speed_str)s|%(progress._eta_str)s',
    '--', url,
  ];

  const proc = spawn(exe, args, { stdio: ['ignore', 'pipe', 'pipe'] });
  _activeProcs.set(jobId, proc);

  const db = getDb();
  const updateProgress = db.prepare('UPDATE jobs SET status=?, progress=?, speed=?, eta=? WHERE id=?');
  const getStatus      = db.prepare('SELECT status FROM jobs WHERE id=?');

  // Poll DB once per second for cancellation
  const cancelTimer = setInterval(() => {
    const row = getStatus.get(jobId);
    if (row?.status === 'cancelled') {
      proc.kill('SIGTERM');
      clearInterval(cancelTimer);
    }
  }, 1000);

  const errorLines = [];

  // yt-dlp sends --progress-template output to stderr in recent versions.
  // Parse progress from both stdout and stderr; collect non-progress lines as errors.
  let buf = '';
  function processChunk(chunk) {
    buf += chunk.toString();
    const lines = buf.split('\n');
    buf = lines.pop();
    for (const line of lines) {
      const m = _PROGRESS_RE.exec(line);
      if (m) {
        const [, status, pct, speed, eta] = m;
        if (status === 'downloading') {
          updateProgress.run('downloading', pct.trim(), speed.trim(), eta.trim(), jobId);
        }
      } else if (line.trim() && !line.startsWith('[')) {
        errorLines.push(line.trim());
      }
    }
  }

  proc.stdout.on('data', processChunk);
  proc.stderr.on('data', processChunk);

  await new Promise(resolve => proc.on('close', resolve));
  clearInterval(cancelTimer);
  _activeProcs.delete(jobId);

  const current = getStatus.get(jobId);
  if (current?.status === 'cancelled') return;

  if (proc.exitCode !== 0) {
    const errMsg = clean(errorLines.join('\n')) || 'Download failed';
    db.prepare("UPDATE jobs SET status='error', error=? WHERE id=?").run(errMsg, jobId);
    return;
  }

  // Find the downloaded file (prefer .mp4, fallback to any file)
  let finalFile = null;
  if (fs.existsSync(outdir)) {
    const files = fs.readdirSync(outdir).filter(f => !f.startsWith('.'));
    if (files.length) {
      const mp4s = files.filter(f => f.toLowerCase().endsWith('.mp4'));
      finalFile  = path.join(outdir, mp4s.length ? mp4s[0] : files[0]);
    }
  }

  if (finalFile) {
    const relPath = `${jobId}/${path.basename(finalFile)}`;
    db.prepare("UPDATE jobs SET status='finished', progress='100%', filename=? WHERE id=?").run(relPath, jobId);
  } else {
    db.prepare("UPDATE jobs SET status='error', error='Output file not found' WHERE id=?").run(jobId);
  }
}

module.exports = { getDownloadOptions, downloadVideo };
